src/plugins/external_tool_support/src/blast_plus/align_worker_subtasks/BlastReadsSubTask.cpp

```diff
@@ -56,21 +56,40 @@ BlastReadsSubTask::BlastReadsSubTask(const QString &dbPath,
       readsNames(readsNames),
       reference(reference),
       minIdentityPercent(minIdentityPercent),
+      readIndex(0),
       storage(storage) {
     setMaxParallelSubtasks(AppContext::getAppSettings()->getAppResourcePool()->getIdealThreadCount());
+    tpm = Progress_Manual;
 }
 
 void BlastReadsSubTask::prepare() {
     QString tempPath = AppContext::getAppSettings()->getUserAppsSettings()->getCurrentProcessTemporaryDirPath();
     CHECK_EXT(!GUrlUtils::containSpaces(tempPath), setError(tr("The task uses a temporary folder to process the data. The folder path is required not to have spaces. "
                                                                "Please set up an appropriate path for the \"Temporary files\" parameter on the \"Directories\" tab of the UGENE Application Settings.")), );
 
-    foreach (const SharedDbiDataHandler &read, reads) {
-        BlastAndSwReadTask *subTask = new BlastAndSwReadTask(dbPath, read, reference, minIdentityPercent, readsNames[read], storage);
-        addSubTask(subTask);
+    // Avoid adding too many subtasks at once. Large numbers of subtasks may work task scheduler slow and freeze the UI.
+    int firstBucketSize = qMin(getMaxParallelSubtasks(), reads.size());
+    for (readIndex = 0; readIndex < firstBucketSize; readIndex++) {
+        const SharedDbiDataHandler &read = reads[readIndex];
+        addSubTask(new BlastAndSwReadTask(dbPath, read, reference, minIdentityPercent, readsNames[read], storage));
+    }
+}
+
+QList<Task *> BlastReadsSubTask::onSubTaskFinished(Task *task) {
+    QList<Task *> newSubtasks;
+    CHECK(!isCanceled(), newSubtasks);
+
+    BlastAndSwReadTask *blastTask = qobject_cast<BlastAndSwReadTask *>(task);
+    SAFE_POINT(blastTask != nullptr, "Must be a BlastAndSwReadTask", newSubtasks);
+    blastSubTasks << blastTask;
```

Comment:
If a blastTask added to the blustSubTasks list on the blastTask finish, then there could be tasks order violation. It is possible in the following case:
1. The first two tasks are created for the corresponding reads and begin execution.
2. The task for the second read finished before the task for the first read and added to the list with index 0.
3. The task for the first read finished and added to the list with index 1.
So these two tasks have incorrect order in the list.

Tasks should be added to blastSubTasks list right after their creation both in prepare and in onSubTaskFinished methods. Of course, progress calculation should be changed properly.

Reply:
Updated the code to preserve the original order.